### DISCUSSION TITLE
응답 형식 확정

### DISCUSSION THREAD

**Discussion by Dr-KoKo** [Strong-Potato/TripVote-BE#13] [General]
Conroller에서 반환할 Dto의 형식에 관한 논의입니다.  
(참고로, #11 에서는 Controller에서 반환할 타입(`ResponseEntity` vs `Dto`)에 관한 논의를 한 바 있습니다.)
제공해주신 몇몇 repo를 확인하고 정리해서 논의사항으로 올리게 되었습니다.

### 1️⃣) status code
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/ed117b04-060d-41bd-86ca-5cd08f9b3cff)
### 2️⃣) message와 응답시간
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/931640b2-679a-494c-93ab-60eaa8af18eb)
### 3️⃣) status code + message
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/3fc9e21b-aaa3-4710-9903-69c4aaebc869)

### 4️⃣) ApiHeader + ApiBody
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/fa1ffda3-b9db-4e6b-99e4-9f8d9364d092)
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/48a83856-f3d6-4f91-a1c9-c57140354745)
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/446a1459-482e-4b8f-85ac-86fe6f19c2bc)
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/e0575ffb-a0c2-4c01-831c-f5cc00e6030f)
### 5️⃣) ApiHeader + ApiBody + ErrorCode
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/80789323-ac06-4a74-bdfd-dd12e8223db6)
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/3042c270-2e6a-472e-96af-3bb31dd21658)
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/b383d73b-016f-4abc-a560-f029263b4459)
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/820eb0cc-af99-461b-98f2-c75bf215f588)
(더 자세한 사항은 https://hongsamm.tistory.com/37를 참고해주세요)

### 6️⃣)정형화된 ApiResponse없이 Dto 반환
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/a0aa29df-00bd-45a5-b3e9-d0e2e1f35709)
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/f62beed8-3e41-41dc-b54b-d641aeb04f23)

**Comment by Dr-KoKo** [comment]
6️⃣번은 프론트에서 code나 message와 같은 메타데이터를 추가요구할 수 있어서 배제하는 것이 좋겠다고 생각합니다. 
4️⃣,5️⃣번도 해본 적은 없는데 메타데이터를 유연하게 바꿀 수 있을 것 같아서 좋을 것 같네요. 다만 `ErrorResponse`는 `RFC7807`을 준수하게 만드려면 `ResponseEntityExceptionHandler`(또는 `ProblemDetailCreator`)를 사용하게 변경해야 할 것 같네요.

**Comment by NoJaeHyuk** [comment]
성공응답에 경우 응답코드(200) 고정 하고 응답바디(ApiResponse)를 정의하여 응답바디만 보고 프런트와 협의하는 방식을 사용하고
공통예외처리에 경우 ResponseEntity + ErrorResponse 형식으로 많이 사용했던것 같습니다. (위 이미지 1~3 패턴) 

4️⃣,5️⃣ 은 한가지 패턴으로 갈 수 있는 장점이 있을거 같습니다.

**Comment by Dr-KoKo** [comment]
에러 핸들링 관련해서 재욱님과 나눈 이야기를 구현 요구사항으로 정리해보았습니다.

## 구현시 요구사항
### 응답형식
- 예외발생시 응답은 `RFC7807`에 따른다.
### Exception 생성시
- App에서 발생시키는 `Exception`은 공통 부모(ex.`ApplicationException`)를 가진다.
- `type`은 'https://tripvote.site/error' 로 한다.(프론트와 논의 후 변경 가능)
- Service에서 `ApplicationException`을 발생시킬 때, 응답할 `title`과 `detail`을 임의로 지정할 수 있다.
### ExceptionHandler
- `ApplicationException`을 핸들링하는 `ExceptionHandler`의 로직이 복잡해서는 안된다.
- `Request`의 파라미터를 `properties`로 반환한다.

이를 토대로 코드를 작성해보았고, 내일 공유드리도록 하겠습니다.

ref)
RFC 공식 문서
 - https://datatracker.ietf.org/doc/html/rfc7807

ResponseEntityExceptionHandler
 - https://velog.io/@appti/ResponseEntityExceptionHandler%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%B4%EC%95%BC-%ED%95%98%EB%8A%94-%EC%9D%B4%EC%9C%A0
 - https://luvstudy.tistory.com/220

Spring Error-Handling
 - https://docs.spring.io/spring-boot/docs/current-SNAPSHOT/reference/html/web.html#web.servlet.spring-mvc.error-handling

Spring ErrerResponse
 - https://docs.spring.io/spring-framework/reference/web/webmvc/mvc-ann-rest-exceptions.html

**Reply by Dr-KoKo** [reply]
에러 핸들링 관련해서 좋은 답변 내용 있어서 올립니다.
[https://www.inflearn.com/questions/111465/5xx-관련-질문있습니다](https://www.inflearn.com/questions/111465/5xx-%EA%B4%80%EB%A0%A8-%EC%A7%88%EB%AC%B8%EC%9E%88%EC%8A%B5%EB%8B%88%EB%8B%A4)

**Comment by laigasus** [comment]
https://github.com/Strong-Potato/TripVote-BE/pull/18 에서 완료되었음
